sdks/java/io/neo4j/src/main/java/org/apache/beam/sdk/io/neo4j/Neo4jIO.java

```diff
@@ -0,0 +1,1221 @@
+        @NonNull SessionConfig sessionConfig,
+        @NonNull TransactionConfig transactionConfig) {
+      this.driverProviderFn = driverProviderFn;
+      this.sessionConfig = sessionConfig;
+      this.transactionConfig = transactionConfig;
+      this.driverSession = DriverSession.emptyClosed();
+    }
+
+    /**
+     * Delay the creation of driver and session until we actually have data to do something with.
+     */
+    @Setup
+    public void setup() {}
+
+    protected @NonNull Driver createDriver() {
+      Driver driver = driverProviderFn.apply(null);
+      if (driver == null) {
+        throw new RuntimeException("null driver given by driver provider");
+      }
+      return driver;
+    }
+
+    protected @Initialized @NonNull DriverSession buildDriverSession() {
+      @NonNull Driver driver = createDriver();
+      @NonNull Session session = driver.session(sessionConfig);
+      return new DriverSession(driver, session);
+    }
+
+    @StartBundle
+    public void startBundle() {
+      if (driverSession == null || driverSession.closed.get()) {
+        driverSession = buildDriverSession();
+      }
+    }
+
+    @FinishBundle
+    public void finishBundle() {
+      cleanUpDriverSession();
+    }
+
+    @Override
+    protected void finalize() {
+      cleanUpDriverSession();
+    }
+
+    protected void cleanUpDriverSession() {
+      if (!driverSession.closed.get()) {
+        try {
+          if (driverSession.session != null) {
+            driverSession.session.close();
+          }
+          if (driverSession.driver != null) {
+            driverSession.driver.close();
+          }
+        } finally {
+          driverSession.closed.set(true);
+        }
+      }
+    }
+
+    protected String getParametersString(Map<String, Object> parametersMap) {
+      StringBuilder parametersString = new StringBuilder();
+      parametersMap
+          .keySet()
+          .forEach(
+              key -> {
+                if (parametersString.length() > 0) {
+                  parametersString.append(',');
+                }
+                parametersString.append(key).append('=');
+                Object value = parametersMap.get(key);
+                if (value == null) {
+                  parametersString.append("<null>");
+                } else {
+                  parametersString.append(value);
+                }
+              });
+      return parametersString.toString();
+    }
+  }
+
+  /** A {@link DoFn} to execute a Cypher query to read from Neo4j. */
+  private static class ReadFn<ParameterT, OutputT> extends ReadWriteFn<ParameterT, OutputT> {
+    protected final @NonNull String cypher;
+    protected final @NonNull RowMapper<OutputT> rowMapper;
+    protected final @Nullable SerializableFunction<ParameterT, Map<String, Object>>
+        parametersFunction;
+
+    private final boolean writeTransaction;
+    private final boolean logCypher;
+
+    private ReadFn(
```

Comment:
Have you considered implementing the Read DoFn as a SplittableDoFn: https://beam.apache.org/documentation/programming-guide/#sdf-basics

Performance of some runners (for example, Dataflow) could significantly suffer when some of the features that SplittableDoFn enable (for example, progress reporting, dynamic work re-balancing) are not supported.